[Jawad280] iP

Jawad280 opened the pull request:
# Tobias Bot 
> "OH MY GOD WE’RE HAVING A FIRE…sale." - Tobias Funke (Arrested Development)

- Text-based
- Easy to learn
- ~~FAST~~ _SUPAH FAST_ to use 

All you need to do is the following : 

1. Download it from [here](https://github.com/nus-cs2103-AY2324S2/ip)
2. Double-click it
3. Add your tasks
4. Let Tobias manage your tasks ! 

Did I mention that it is **ABSOLUTELY FREE!!!**

Features:
- [x] Track your tasks
- [x] Have wonderful conversations
- [ ] Bring him to real life

### Some code for yall noobs

If you are a Java God, you can use it to practice Java too. Here's the `main` method:
```
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Jawad280 committed:
Level 0 added

Jawad280 committed:
level-1 Echo updated

Jawad280 committed:
Level 2 : Added list

Jawad280 committed:
Level 3 : Mark as done

Jawad280 committed:
Level 4 : Todos, Event, Deadlines

Jawad280 committed:
accidentally ignored the external classes, adding them back

Jawad280 committed:
Renamed Duke, Added TobiasExceptions and Delete function

Jawad280 committed:
Simplified taskPrinter() & added a checkmark

Jawad280 committed:
A-TextUiTesting added

Jawad280 committed:
Level-7

Jawad280 committed:
Level 8 - Handling DateTime Objects

Jawad280 committed:
A-MoreOOP

Jawad280 committed:
A-Packages : Partitioned the classes into respective packages

Jawad280 committed:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631' into branch-Level-8

* commit 'a6f73244a20539e20c2189973ebf597df34ef631':
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

Jawad280 committed:
A-Gradle & A-JUnit & A-Jar

Jawad280 committed:
Add JavaDocs

Jawad280 committed:
A-CodingStandard updated

Jawad280 committed:
Search functionality added

Jawad280 committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

* branch-A-JavaDoc:
  Add JavaDocs

Jawad280 committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

* branch-A-CodingStandard:
  A-CodingStandard updated

Jawad280 committed:
Resolved conflicts : Javadocs missing and some coding standards not consistent with branch-9

Jawad280 committed:
Adding tasks

Jawad280 committed:
Update Commands

Jawad280 committed:
Add Javadocs to Storage.java

Jawad280 committed:
Format Parser.java

Reformatted the Parser functions and fleshed out certain functionalities. Added functionality of find.

Jawad280 committed:
Add JavaDocs to NoreplyCommand and other classes.

Some minor modifications to the formatting of code to adhere to Coding Standards

Jawad280 committed:
Revert "Add JavaDocs to NoreplyCommand and other classes."

This reverts commit 81764c987c1d73a3c3bf7c30f1aa1cb13b183a4e.

Jawad280 committed:
Update Ui.java

Reformatted the Ui class to adhere to java coding standards.
Customised the replies given by the bot.

Jawad280 committed:
Update TaskList.java

Reformat the printListFind(keyword) function to better adhere to the
coding standards. Added some custom log messages to better showcase
the functionality of the printListFind(keyword) method

Jawad280 committed:
Update Storage.java

Upon creating the proper dirs and files for the save file, added a new
log message to indicate if that action took place.
If there is already a save file -> Existing message
If there is no save file -> Creating dir message

Jawad280 committed:
Update Storage.java

Included JavaDocs for all the functions.
More inclusions of log messages such as when the file is corrupted.

Jawad280 committed:
Update FindCommand

JavaDocs for the latest Command (Find) is added to give better
understanding of the execute() function.
These docs will help to quickly identify the purpose of each function
when we need to edit or update them in the future.

Jawad280 committed:
Update Tobias.java

Included JavaDocs to the constructor for Tobias and to all its
relevant functions. This will help to deduce quickly what each function
does in the future.

Jawad280 committed:
Update existing classes and include javaFX

By adding javaFX, removed the command line responses and replaced it
with Strings which is then printed onto the GUI of Tobias.

lihaoquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,139 @@
+            boolean isDone = Integer.parseInt(data.substring(8,9)) == 1;
+
+            int desc = data.indexOf("|desc");
+            String description = data.substring(desc+5);
```
As per the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html) there should be a space between operators.
```suggestion
            String description = data.substring(desc + 5);
```

lihaoquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,139 @@
+
+            int desc = data.indexOf("|desc");
+            int by = data.indexOf("|by");
+            String description = data.substring(desc+5, by);
+            String byDate = data.substring(by+3);
```
```suggestion
            String description = data.substring(desc + 5, by);
            String byDate = data.substring(by + 3);
```

lihaoquan commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,139 @@
+            int desc = data.indexOf("|desc");
+            int from = data.indexOf("|from");
+            int to = data.indexOf("|to");
+            String description = data.substring(desc+5, from);
+            String fromDate = data.substring(from+5, to);
+            String toDate = data.substring(to+3);
```
```suggestion
            String description = data.substring(desc + 5, from);
            String fromDate = data.substring(from + 5, to);
            String toDate = data.substring(to + 3);
```

lihaoquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+            }
+
+            return new DeadlineCommand(description, dd);
+        }
+        else if (command.startsWith("event")) {
```
Referring to the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html) under 

>  The if-else class of statements should have the following form:

```suggestion
        } else if (command.startsWith("event")) {
```

lihaoquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+            LocalDateTime t = dateFromString(to);
+
+            return new EventCommand(description, f, t);
+        }
+        else if (command.startsWith("mark")) {
```
Referring to the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html) under 

>  The if-else class of statements should have the following form:

```suggestion
        } else if (command.startsWith("mark")) {
```

lihaoquan replied:
There are some similar structures below as well.

lihaoquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+package parser;
+
+import command.*;
```
Referring to the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html) under 

>   Imported classes should always be listed explicitly.

Wildcard import is listed under "Bad" example.

lihaoquan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+public class Parser {
+
+    
+    /**
+     * Takes in a string representation of date/time and returns it as a LocalDateTime object.
+     *
+     * @param dateTime String of the required date/time information.
+     * @return a LocalDateTime object with the provided date/time information.
+     * @throws TobiasException if the dateTime string does not follow format of dd-MM-yyyy HHmm.
+     * */
```
I like the thorough effort in the Javadocs 👍

lihaoquan commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,9 @@
+import task.TaskList;
+import ui.Ui;
+
+public abstract class Command {
+    public abstract String execute(TaskList tasks, Ui ui, Storage storage);
+    public abstract boolean isExit();
+}
```
Nice use of abstract parent class to manage the commands.

lihaoquan commented on src/main/java/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,66 @@
+     * @param ui Current Ui.
+     * */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        Task task = new Deadline(description, isDone, deadline);
+        tasks.addToList(task);
+
+        return ui.addedTaskPrinter(task, tasks.taskNum());
+    }
```
Is this perhaps Strategy Pattern? Nice 👍

lihaoquan reviewed (commented):
Generally, it looks great, besides coding standard errors.
Checkstyle can identify these errors thoroughly.

I see some nice programming design patterns that I can learn from, in particular for managing commands.

ejnan commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,89 @@
+    private Tobias tobias;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/michaelUser.jpeg"));
+    private Image tobiasImage = new Image(this.getClass().getResourceAsStream("/images/tobiasBot.jpeg"));
```
Should `userImage` and `tobiasImage` be set as constants? They seem to represent fix images in this case.

ejnan commented on src/main/java/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,66 @@
+     * @param ui Current Ui.
+     * */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
Maybe a more descriptive method name to enhance clarity and readability? 
`executeDeadlineCommand` for example?

ejnan commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,52 @@
+     * @param ui Current Ui.
+     * */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
```
Maybe a more descriptive method name to enhance clarity and readability?
`executeDeleteCommand` for example?

ejnan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+            } else {
+                throw new TobiasException("Invalid number provided! Please give a valid index from 1 to " + size +" !!");
+            }
+        }
+        else if (command.startsWith("unmark")) {
```
Maybe you would want to have else if statement on the same line as the closing brackets from the previous statement? This would follow our java coding standard:

```
if (condition) {
    statements;
} else if (condition) {
    statements;
} else {
    statements;
}
```

ejnan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+            } else {
+                throw new TobiasException("Invalid number provided! Please give a valid index from 1 to " + size +" !!");
+            }
+        }
+        else if (command.startsWith("delete")) {
```
Again, maybe you would want to have else if statement on the same line as the closing brackets from the previous statement? This would follow our java coding standard:

```
if (condition) {
    statements;
} else if (condition) {
    statements;
} else {
    statements;
}
```

ejnan commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,244 @@
+
+            String keyWord = command.substring(5);
+            return new FindCommand(keyWord);
+        }
+        else if (command.equals("bye")) {
+            return new ExitCommand();
+        }
+        else {
```
Again, maybe you would want to have else if and else statement on the same line as the closing brackets from the previous statement? This would follow our java coding standard:

```
if (condition) {
    statements;
} else if (condition) {
    statements;
} else {
    statements;
}
```

ejnan reviewed (commented):
Overall, I found your code easy to read - following the coding standard and naming conventions for the most part. I like how you separated each class to handle one thing each. I noted a few minor coding standard violations; something you could consider to change for the execute method for all of the commands. Other than that, LGTM!

Jawad280 committed:
Modify code to adhere to code quality

There were 2 instances of improper else-if code format that
deviates from standard code quality.

Clean up Tobias.java to remove unused function, run(), which was
initially used in the CLI interface

Jawad280 committed:
Merge pull request #1 from Jawad280/A-CodeQuality

Modify code to adhere to code quality

Jawad280 committed:
Assertions to Ui class

These assertions check for the following :
- Task list is not empty or size is not negative
- Constants are not null (ie not altered)

Jawad280 committed:
Merge pull request #2 from Jawad280/A-Assertions

Assertions to Ui class

Jawad280 committed:
Tagging is incorporated

Tag - Add tags to a task
Untag - Removes a tag from a task

Jawad280 committed:
Merge pull request #3 from Jawad280/C-Tagging

Tagging is incorporated

Jawad280 committed:
Update Ui message

By modifying the custom messages sent by the bot
we can make it sound more unique.

Moving forward, we can incorporate more custom commands
and custom replies.

Jawad280 committed:
Update Command

With the inclusion of a ReplyCommand, one can add custom messages
to be replied to when a command is given.

This will help improve the quality of the bot's replies.

Jawad280 committed:
Update Parser

The TobiasException message will now reflect what the user
is able to ask the bot.

Jawad280 committed:
Update TagCommand, UntagCommand, ReplyCommand

Tag, Untag and Reply Commands has Java Docs.

Java Docs makes it easier to explain the functionalities of
the Tag, Untag and Reply Commands.

Jawad280 committed:
Update MainWindow

Moving command guide from Parser to a Popup in the main window
makes it easier and clutter-free for user to view the user-guide.

benson1029 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,141 @@
+        Alert aboutPopup = new Alert(Alert.AlertType.INFORMATION);
+        aboutPopup.setTitle("Commands");
+        aboutPopup.setHeaderText("Tobias Bot");
+        String info = "You can ask me to do the following quests for you, mate:" +
```
This method is too long. Consider isolating the magic constant `info`.

benson1029 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private void showWelcomeMessage() {
+        String result = Ui.printDivider()
```
weak SLAP, as you're combining results from higher-level methods with constant literal strings.

benson1029 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,307 @@
+                throw new TobiasException("Hey, please enter a description !");
+            }
+
+            String blank = command.substring(4,5);
```
weak SLAP, you should isolate the command processing (especially finding substrings) from the matching of commands

benson1029 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,307 @@
+                throw new TobiasException("Hey, please enter a /from date/day/time!!");
+            }
+
+            if (toIndex == -1) {
```
This method is too long. Isolate command format checking from this matching stage. Perhaps make a method for each of the commands with non-trivial parameters?

benson1029 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,151 @@
+        try {
+            if (tasks.isEmpty()) {
+                throw new TobiasException("Your list is empty at the moment !");
+            } else {
```
Nesting problems. Don't put the default control flow in the else branch.

benson1029 reviewed (commented):
Code is neat overall. However, consider fixing the instances of weak SLAP, too long methods and nested code (where the happy path is not prominent).

Jawad280 committed:
Update Parser.java with SLAP

Initially the parser had a very long if-else block with
all the conditions. By separating the conditions into a function
on its own, we can achieve better level-abstraction.

Jawad280 committed:
Update multiple import lines

Initially, there was usage of wildcard imports in a few
of the classes. Rectifying it by only choosing what we need to
import.

Jawad280 committed:
Update build and add new Command

The SaveCommand allows simply to save the list to the local
data. A few changes made to the JavaFX to make the GUI more
appealing. Inclusion of a proper User Guide in the help
menu button.

Jawad280 committed:
Update MainWindow.fxml for name

Inclusion of the name of the bot in the menu bar of the app.
The label is added at the top most layer of the menu bar.

Jawad280 committed:
Add Ui.png to the docs

A screenshot of the Ui of Tobias Bot that is captured to
showcase the functionalities of the bot. It has been populated
with realistic data.

Jawad280 committed:
Update README.md

Gives a quick overview of Tobias bot.
It provides all the necessary features and commands that the
user can use and also its related outcomes.

Jawad280 committed:
Update readme.md

Gives a quick overview of Tobias bot.
It provides all the necessary features and commands that the
user can use and also its related outcomes.

Jawad280 committed:
Update readme.md

Gives a quick overview of Tobias bot.
It provides all the necessary features and commands that the
user can use and also its related outcomes.
Edit some typos for clarity.

Jawad280 committed:
Update SLAP and readme

SLAP was further implemented in Storage and Parser to avoid
use of long methods and deep nesting. Readme.md was updated to
include the function of find as well.

Jawad280 committed:
Update MainWindow.java

Instead of printing the user guide as a long string, it is now
a hyperlink to the User guide website on github.io.

Jawad280 committed:
Update README.md

Changes made to the readme to resolve some inconsistencies and
fix typos as well as overall outline of the UG.
Inclusion of the tag A-UserGuide.

Jawad280 committed:
Update README.md

Changes made to the readme to resolve some inconsistencies and
fix typos as well as overall outline of the UG.
Inclusion of the tag A-UserGuide.

Jawad280 committed:
Update Commands

Initially all executions of the command will not save the tasklist.
Task List will only save if the save or by command was called. Now
it will automatically save the tasklist.
